### PR TITLE
inrepoconfig: retire secondary clone serialization for same-repo requests

### PR DESCRIPTION
**Pull request by listx**
Before 916ed2b4b7 (Merge pull request #30400 from
listx/inrepoconfig-optimizations, 2023-08-17), we were under the
impression that the secondary clones of inrepoconfig repos were
unavoidably large. So we created the InRepoConfigGitCache wrapper in
3098b88178 (Merge pull request #23264 from cjwagner/inrepo-config-cache,
2021-08-18) to make it so we would never delete and recreate the
secondary clone for every request (the default behavior of the wrapped
git v2 client); instead, we would reuse the existing secondary clones
across requests if those requests were for the same repo. This resulted
in the need to ensure that only one such request would touch the
associated secondary clone at any given time, which meant that we had to
serialize the use of these secondary clones.

It turns out that the secondary clone optimizations in 916ed2b4b7 make
these clones cheaper on disk by multiple orders of magnitude. In manual
tests, we saw reductions in size from 3.1GB to just 19MB for the disk
space used for a secondary clone for a large repo using inrepoconfigs
(where the bulk of the 19MB came from other unrelated files in the
toplevel folder of the repo, due to the way sparse checkouts work).

Because the secondary clones are so cheap now, there's no need to try to
reuse the same one during the lifetime of a process. We just create and
delete each time we need to service a request.

Note that we already have a call to repo.Clean() inside prowYAMLGetter,
so that will delete the repo after we read the inrepoconfig values.
However, the main "gitcache" directory (holding the mirror clone) will
still be around and will not be deleted.

/cc @cjwagner @airbornepony @timwangmusic

### CODE REVIEW

File: prow/git/v2/interactor_test.go
Comment (airbornepony) [inline prow/git/v2/interactor_test.go:110]: If we use the strings "/dir" and "/from", it's easier to observe the positional behavior in the expected calls.

File: prow/config/inrepoconfig.go
Comment (airbornepony) [inline prow/config/inrepoconfig.go]: Use inRepoConfigDirName instead of the literal?

File: prow/git/v2/interactor.go
Comment (airbornepony) [inline prow/git/v2/interactor.go:152]: Could simplify the API surface by adding a variadic (optional) CloneOpts arg to the existing Clone().

File: prow/git/v2/client_factory.go
Comment (cjwagner) [inline prow/git/v2/client_factory.go:267]: Was this just for debugging or did you mean to include this? These seem like they might be a bit spammy for the info level, but it doesn't matter much.

**Review by cjwagner — COMMENTED** [review]
I just reviewed the last commit. LGTM.

File: prow/git/v2/client_factory.go
Comment (listx) [reply prow/git/v2/client_factory.go:267]: Oh yeah these were for my own debugging. Will clean it up.

**Comment by listx** [comment]
/hold

The tests are passing but I want to clean up the commit message and PR description first.

**Comment by listx** [comment]
Somehow, deleting the (now unnecessary) concurrency test in inrepoconfig_test.go has made the other unrelated race detections (for tests in config_test.go, not even in inrepoconfig_test.go) go away.

File: prow/gerrit/adapter/adapter_test.go
Comment (airbornepony) [inline prow/gerrit/adapter/adapter_test.go:562]: Call above should be one line.

**Review by airbornepony — APPROVED** [review]

**Review by cjwagner — APPROVED** [review]

File: prow/gerrit/adapter/adapter_test.go
Comment (listx) [reply prow/gerrit/adapter/adapter_test.go:562]: Just saw this. Cole already approved and I don't want to ask him to re-LGTM for such a small change. Will queue in my cleanup followups branch, thanks.

Comment (listx) [reply prow/gerrit/adapter/adapter_test.go:562]: See commit "style: prefer single line for function arguments" at #30428.

**Pull request merged**
